File: typescript/src/__tests__/config.test.ts
import getConfig, { Operation } from "../config";

test("simple test print", function() {
    const config = getConfig({});
    expect(config.operation).toEqual(Operation.Print);
    expect(config.args).toEqual([]);
});
test("simple test print with argument 1", function() {
    const config = getConfig({ args: ["foo"] });
    expect(config.operation).toEqual(Operation.Print);
    expect(config.args).toEqual(["foo"]);
});

test("simple test add", function() {
    const config = getConfig({
        args: ["add", "foo", "bar"]
    });
    expect(config.operation).toEqual(Operation.Add);
    expect(config.args).toEqual(["foo", "bar"]);
});

test("simple test remove", function() {
    const config = getConfig({
        args: ["rm", "foo"]
    });
    expect(config.operation).toEqual(Operation.Remove);
    expect(config.args).toEqual(["foo"]);
});

File: go/pkg/projector/config_test.go
package projector_test

import (
	"reflect"
	"testing"

	"github.com/RyanDev-21/pkg/projector"
)

func getOpts(args []string) *projector.Opts {
	return &projector.Opts{
		Args:   args,
		Pwd:    "",
		Config: "",
	}
}

func testConfig(t *testing.T, args []string, expected []string, op projector.Operation) {
	opts := getOpts(args)
	config, err := projector.NewConfig(opts)
	if err != nil {
		t.Errorf("expected to get no error: %v", err)
	}
	if !reflect.DeepEqual(config.Operation, op) {
		t.Errorf("expected operation to be %v :got %v", op, config.Operation)
	}
	// if !reflect.DeepEqual(config.Config, configPath) {
	// 	t.Errorf("expected operation to be %v :got %v", configPath, config.Config)
	// }
	// if !reflect.DeepEqual(config.Pwd, pwd) {
	// 	t.Errorf("expected operation to be %v :got %v", pwd, config.Pwd)
	// }
	switch op {
	case projector.Print:
		if !reflect.DeepEqual(config.Args, args) {
			t.Errorf("expected args to be %v :got %v", expected, config.Args)
		}
	default:
		if !reflect.DeepEqual(config.Args, args[1:]) {
			t.Errorf("expected args to be %v :got %v", expected, config.Args)
		}
	}
}

func TestConfigPrintKey(t *testing.T) {
	testConfig(t, []string{}, []string{}, projector.Print)
}

func TestConfigPrintKeyValue(t *testing.T) {
	testConfig(t, []string{"foo"}, []string{"foo"}, projector.Print)
}

func TestConfigAddKey(t *testing.T) {
	testConfig(t, []string{"add", "foo", "bar"}, []string{"foo", "bar"}, projector.Add)
}

func TestConfigRemoveKey(t *testing.T) {
	testConfig(t, []string{"rm", "foo"}, []string{"foo"}, projector.Remove)
}
